File: go/auth_scope_mint_test.go
package realmid

import (
	"context"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"testing"
)

// ---- ADR-097 mint half: TokenRequest.Scope ----
//
// The enforcement half of ADR-097 (ScopesFrom / ScopeAllows / ScopePolicy)
// shipped in all three SDKs; the MINT half shipped in none of them, so
// `ScopePolicy` was reachable only by a partner who bypassed the SDK and
// called POST /auth/token by hand. These tests lock the operand onto the wire.

// TestAuth_TokenSendsScope is the defect test: before this field existed the
// body had no `scope` key at all, whatever the caller asked for.
func TestAuth_TokenSendsScope(t *testing.T) {
	var gotBody map[string]any
	srv := authTestServer(t, map[string]http.HandlerFunc{
		"/auth/token": func(w http.ResponseWriter, r *http.Request) {
			buf, _ := io.ReadAll(r.Body)
			_ = json.Unmarshal(buf, &gotBody)
			_ = json.NewEncoder(w).Encode(map[string]any{
				"access_token": "atok", "refresh_token": "rtok", "expires_in": 900,
			})
		},
	})
	defer srv.Close()

	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
	if _, err := r.Auth.Token(context.Background(), TokenRequest{
		RefreshToken: "rtok",
		TenantID:     "t1",
		Scope:        []string{"orders:read", "orders:write"},
	}); err != nil {
		t.Fatalf("token: %v", err)
	}
	// Space-delimited, in the order given — RFC 6749 §3.3. Order is preserved
	// because ScopesFrom promises the issuer's order back to the verifier.
	if got, _ := gotBody["scope"].(string); got != "orders:read orders:write" {
		t.Errorf("scope on the wire = %q, want %q (body: %+v)", got, "orders:read orders:write", gotBody)
	}
}

// TestAuth_TokenOmitsAbsentScope keys omission on emptiness, NOT on nil.
//
// This is deliberately DIFFERENT from RolePermissions, which is nil-keyed
// because an empty supplied list is a real instruction there ("this role
// confers nothing here", answered with a 403). An empty `scope` has no such
// meaning: the issuer's parseScope trims and returns nil for "", so sending
// `"scope": ""` and omitting the key are indistinguishable server-side.
// Keying on nil would put a field on the wire that cannot mean anything.
func TestAuth_TokenOmitsAbsentScope(t *testing.T) {
	for _, tc := range []struct {
		name  string
		scope []string
	}{
		{"nil", nil},
		{"empty", []string{}},
	} {
		t.Run(tc.name, func(t *testing.T) {
			var gotBody map[string]any
			srv := authTestServer(t, map[string]http.HandlerFunc{
				"/auth/token": func(w http.ResponseWriter, r *http.Request) {
					buf, _ := io.ReadAll(r.Body)
					_ = json.Unmarshal(buf, &gotBody)
					_ = json.NewEncoder(w).Encode(map[string]any{
						"access_token": "atok", "refresh_token": "rtok", "expires_in": 900,
					})
				},
			})
			defer srv.Close()

			r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
			if _, err := r.Auth.Token(context.Background(), TokenRequest{
				RefreshToken: "rtok", TenantID: "t1", Scope: tc.scope,
			}); err != nil {
				t.Fatalf("token: %v", err)
			}
			if _, present := gotBody["scope"]; present {
				t.Errorf("scope key present for a %s scope: %+v", tc.name, gotBody)
			}
		})
	}
}

// TestAuth_TokenRefusesUnsendableScope is the reason Scope is []string and not
// a raw string.
//
// A SPACE inside one entry is not a parse error on the wire — it is a SILENT
// AUTHORITY CHANGE. `[]string{"orders read"}` joins to "orders read", which the
// issuer's strings.Fields reads as TWO scopes, minting authority the caller
// never asked for. Taking a list and refusing an unsendable entry client-side
// turns that into an error the partner sees at the call site.
//
// The charset is RFC 6749 §3.3 and is FIXED BY SPEC, not per-realm — which is
// what makes checking it here safe from drift. The per-realm bounds
// (max_permission_strings / max_permission_string_len) are deliberately NOT
// checked here: those ARE realm-configurable, so a client-side copy would
// refuse tokens the server would have accepted.
func TestAuth_TokenRefusesUnsendableScope(t *testing.T) {
	for _, tc := range []struct {
		name  string
		entry string
	}{
		{"embedded space splits into two scopes", "orders read"},
		{"tab is whitespace the issuer also splits on", "orders\tread"},
		{"double quote is outside the scope-token charset", `orders"read`},
		{"backslash is outside the scope-token charset", `orders\read`},
		{"empty entry cannot be represented at all", ""},
		{"del is not printable ASCII", "orders\x7f"},
	} {
		t.Run(tc.name, func(t *testing.T) {
			reached := false
			srv := authTestServer(t, map[string]http.HandlerFunc{
				"/auth/token": func(w http.ResponseWriter, r *http.Request) {
					reached = true
					_ = json.NewEncoder(w).Encode(map[string]any{"access_token": "atok"})
				},
			})
			defer srv.Close()

			r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: srv.URL})
			_, err := r.Auth.Token(context.Background(), TokenRequest{
				RefreshToken: "rtok", TenantID: "t1",
				Scope: []string{"orders:read", tc.entry},
			})
			if !errors.Is(err, ErrInvalidScope) {
				t.Fatalf("err = %v, want ErrInvalidScope", err)
			}
			// The mint must not happen at all — a refusal that still spends the
			// refresh token would rotate it and log the caller out.
			if reached {
				t.Error("request reached the issuer; the refusal must be client-side")
			}
		})
	}
}

File: ts/src/auth-scope-mint.test.ts
import { test } from "node:test";
import { strict as assert } from "node:assert";
import { createRealm } from "./realm.js";
import { RealmError } from "./errors.js";

// ---- ADR-097 mint half: TokenRequest.scope ----
//
// scope.ts is the ENFORCEMENT half — scopesFrom / scopeAllows / scopePolicy —
// and it shipped in all three SDKs. The MINT half shipped in none of them, so
// the operand those functions evaluate had no way onto the wire from the SDK
// at all. These tests hold it there.

interface Recorded { url: string; body: unknown }

/** Auto-services the platform-token bootstrap, records everything else. */
function recorder(handler: () => Response): { fetch: typeof fetch; calls: Recorded[] } {
  const calls: Recorded[] = [];
  const fetch = (async (input: RequestInfo | URL, init?: RequestInit): Promise<Response> => {
    const url = typeof input === "string" ? input : input.toString();
    let body: unknown;
    if (init?.body) { try { body = JSON.parse(String(init.body)); } catch { body = init.body; } }
    if (url.endsWith("/auth/login") && (body as { grant_type?: string })?.grant_type === "platform_api_key") {
      return new Response(JSON.stringify({
        status: "ok", subject_type: "platform", refresh_token: "rtok-platform",
        access_token: "pt_test_abc", expires_in: 300,
      }), { status: 200, headers: { "content-type": "application/json" } });
    }
    // Only the mint is recorded. The platform-token bootstrap and the
    // GET /platforms/mine that follows it are harness traffic, and counting
    // them would make `calls.length === 0` assert something else entirely.
    if (url.endsWith("/auth/token")) calls.push({ url, body });
    return handler();
  }) as typeof fetch;
  return { fetch, calls };
}

const okMint = () => new Response(JSON.stringify({
  access_token: "at2", refresh_token: "rt2", expires_in: 900, subject_type: "user",
}), { status: 200, headers: { "content-type": "application/json" } });

const REALM_ID = "01HREALM";
const API_KEY = "rk_live_test123";
const mk = (fetch: typeof fetch) =>
  createRealm({ realmId: REALM_ID, apiKey: API_KEY, baseUrl: "https://auth.test", fetch });

// The defect test: before this field existed the body had no `scope` key,
// whatever the caller asked for.
test("auth.token: scope goes on the wire space-delimited, in order", async () => {
  const { fetch, calls } = recorder(okMint);
  await mk(fetch).auth.token({
    refreshToken: "rt", tenantId: "t1", scope: ["orders:read", "orders:write"],
  });
  const body = calls[0]!.body as Record<string, unknown>;
  assert.equal(body.scope, "orders:read orders:write");
});

// Keyed on emptiness, NOT on undefined — the inverse of rolePermissions, and
// for a stated reason: the issuer's parseScope trims and returns nil for "",
// so an empty scope IS an absent one and `"scope": ""` could not mean anything.
// rolePermissions differs because an empty list there is a real instruction
// ("this role confers nothing here"), answered with a 403.
for (const [name, scope] of [
  ["undefined", undefined],
  ["empty", [] as string[]],
] as const) {
  test(`auth.token: omits the scope key entirely for a ${name} scope`, async () => {
    const { fetch, calls } = recorder(okMint);
    await mk(fetch).auth.token({ refreshToken: "rt", tenantId: "t1", scope });
    const body = calls[0]!.body as Record<string, unknown>;
    assert.equal("scope" in body, false, `scope key present: ${JSON.stringify(body)}`);
  });
}

// The reason scope is string[] and not the wire's raw string.
//
// A SPACE inside one entry is not a parse error on the wire — it is a SILENT
// AUTHORITY CHANGE: "orders read" is read by the issuer as TWO scopes. Taking a
// list and refusing an unsendable entry turns that into an error at the call
// site. The charset is RFC 6749 §3.3 and fixed by spec, which is what makes
// checking it client-side safe from drift; the per-realm BOUNDS
// (max_permission_strings / max_permission_string_len) are deliberately left to
// the server, because those are realm configuration and a local copy would
// refuse what the server accepts.
for (const [name, entry] of [
  ["an embedded space, which splits it into two scopes", "orders read"],
  ["a tab, which the issuer also splits on", "orders\tread"],
  ['a double quote, outside the scope-token charset', 'orders"read'],
  ["a backslash, outside the scope-token charset", "orders\\read"],
  ["nothing at all", ""],
  ["DEL, which is not printable ASCII", "orders\x7f"],
] as const) {
  test(`auth.token: refuses a scope entry containing ${name}`, async () => {
    const { fetch, calls } = recorder(okMint);
    await assert.rejects(
      () => mk(fetch).auth.token({
        refreshToken: "rt", tenantId: "t1", scope: ["orders:read", entry],
      }),
      (e: unknown) => e instanceof RealmError && e.code === "bad_request",
    );
    // The mint must not happen at all — a refusal that still spent the refresh
    // token would rotate it and log the caller out.
    assert.equal(calls.length, 0, "the request must not reach the issuer");
  });
}
